### PR TITLE
Better structure the KV cache, make it mutable, non-optional

### PR DESCRIPTION
**Pull request by danieldk**
Before this change, the KV cache was passed around as `Option<Vec<KeyValueCache>>`. This has several issues. First, the use of an option requires a lot of wrapping and unwrapping. Second, the cache is immutable and in the future we want to pre-allocate the cache and mutate it in-place to avoid allocations/copies.

This change removes the `Option` wrapping and makes the cache mutable. We don't do in-place updates yet.

The type for the cache itself is now `KeyValueCache` and the cache for a layer is `LayerKeyValueCache`. Decoders now always require passing a `KeyValueCache` object. There is a special constructor, `KeyValueCache::no_cache` that creates a no-op cache.

This approach allows us to treat the cache/no-cache cases in the same way. The same operations are applied, but they are just ignored when `no_cache` is used.

### CODE REVIEW

File: src/kv_cache.rs [hunk 1]
```diff
@@ -1,10 +1,203 @@
...
-    pub attention_mask: AttentionMask,
+pub struct KeyValueCache(KeyValueCacheEnum);
+
+impl KeyValueCache {
+    /// Create a key-value cache.
+    ///
+    /// * `batch_size` - Batch size.
+    /// * `hidden_width` - Hidden width.
+    /// * `n_key_value_heads` - Number of key-value heads.
+    /// * `n_hidden_layers` - Number of hidden layers.
```
Comment (shadeMe) [inline src/kv_cache.rs]:
```suggestion
    /// * `n_layers` - Number of hidden layers.
```

File: src/layers/attention/self_attention.rs [hunk 2]
```diff
@@ -387,38 +387,28 @@ impl Attention for SelfAttention {
...
-                cache
-                    .attention_mask
-                    .extend(attention_mask)
-                    .context(AttentionMaskSnafu)?,
-            ),
-            None => (key, value, attention_mask.clone()),
-        };
+        cache.update(&key, &value)?;
+        let key = cache.key().unwrap_or(&key);
+        let value = cache.value().unwrap_or(&value);
```
Comment (shadeMe) [inline src/layers/attention/self_attention.rs:392]: Looks like we don't cache the attention mask anymore? Doesn't look like it's intentional?

**Review by shadeMe — CHANGES_REQUESTED** [review]

File: src/layers/attention/self_attention.rs (hunk 2 above)
Comment (danieldk) [reply src/layers/attention/self_attention.rs:392]: I should have put this in the commit message. It's intentional, because I am not sure where it should go. Before we were doing a lot of unnecessary allocations because we stored an attention mask for every layer.

But if you store it only once in the cache, who is responsible for updating the mask? Not a particular layer, because then other layers need to know that they are not the first layer. I guess it could be done in the decoder, but then the decoder and decoder layer interfaces would diverge - the decoder would only expect an attention mask for the new tokens and the layers for all tokens.

Comment (danieldk) [reply src/layers/attention/self_attention.rs:392]: Updated the commit message with more information why the mask is removed from the cache for the time being.

Comment (shadeMe) [reply src/layers/attention/self_attention.rs:392]: Makes sense.

**Review by shadeMe — APPROVED** [review]
